assets/javascripts/initializers/code-bytes.js

```diff
@@ -175,8 +176,36 @@ function initializeCodeByte(api) {
         div.appendChild(saveButton);
       }
     });
-  }),
-    { id: 'codebyte-preview' };
+  },
+  { id: 'codebyte-preview' });
+
+  api.modifyClass("controller:composer", {
+    save(...args) {
+      let allCodebytesAreValid = true;
+      let index = 0;
+      let start, end;
+      const inputLines = this.model.reply.split('\n');
+
+      do {
+        [start, end] = findCodeByte(inputLines, index);
+        index++;
+        if (start !== undefined && !inputLines[start].match(/^\[codebyte[ ]+language=([^\s]*?)[ ]*]$/)) {
```

Comment:
would you mind putting this regex in a const and sharing it with `findCodeByte`? the only thing scarier than one regex is two of them

Reply:
also this has the same problem that `findCodeByte` has, which is that `[codebyte language=]` is considered valid. At the time I wasn't sure how to make what comes after the `=` one or more characters that _aren't_ the closing `]`, but maybe something like

`/^\[codebyte[ ]+language=([^\s^\]]+)[ ]*]$/`

would work? If not it's not really a big deal, with this change it only affects people who are writing the tags by hand for some reason

Reply:
They are actually two different regexes. The one in `findCodeByte` `/^\[codebyte([ ]+language=([^\s]*?))?[ ]*]$/` matches a codebyte open tag with or without that language attribute, as we want to find new codebytes that don't have a language yet. The one here `/^\[codebyte[ ]+language=([^\s]*?)[ ]*]$/` is similar but only matches a codebyte open tag with the language attribute. So, [codebyte] matches the first regex but not the second, which is how we tell that it's in an invalid state.

That said, would it be better to put both of these in constants at the top of the file? That might make the differences clearer.

Reply:
gotcha, I think commenting one or both would be best actually

Reply:
I made it so `[codebyte language=]` is no longer considered valid for this save logic.

For `findCodeByte`, I don't think we need to check the language attribute at all. So, I removed that part from the regex altogether. Previously, `findCodeByte` wouldn't match tags with a non-language attribute, e.g. `[codebyte language=csharp foo = bar]` or `[codebyte lang=javascript]`. But, Discourse did treat those as valid and would render them. So, I think our logic should match that, otherwise it would be confusing as to why some codebytes couldn't be saved. The other attributes will get cleared out on save anyway.

@christian-dinh What do you think about that change? Is there a reason that `findCodeByte` needs to check the language attribute?

Reply:
yeah I think it's fair to say that if discourse renders the tag it should be considered valid, so this is fine with me!